onnxruntime/python/tools/kernel_explorer/kernels/gemm_ck.cc

```diff
@@ -0,0 +1,188 @@
+      arg_ = typed_impl->MakeArgumentPointer(this->a_, this->b_, this->c_,
+                                             this->m_, this->n_, this->k_,
+                                             this->lda_, this->ldb_, this->ldc_,
+                                             nop, nop, nop);
+      dispatch_ = typed_impl->MakeInvokerPointer();
+    } else if (this->opa_ == BlasOp::T && this->opb_ == BlasOp::T) {
+      using DeviceOp = DeviceGemm<Col, Col>;
+      auto typed_impl = static_cast<DeviceOp*>(impls_[selected_impl_].get());
+      arg_ = typed_impl->MakeArgumentPointer(this->a_, this->b_, this->c_,
+                                             this->m_, this->n_, this->k_,
+                                             this->lda_, this->ldb_, this->ldc_,
+                                             nop, nop, nop);
+      dispatch_ = typed_impl->MakeInvokerPointer();
+    }
+
+    return impls_[selected_impl_]->IsSupportedArgument(arg_.get());
+  }
+
+ public:
+  CKGemm(BlasOp opa, BlasOp opb,
+         int64_t m, int64_t n, int64_t k,
+         double alpha,
+         DeviceArray& a, int64_t lda,
+         DeviceArray& b, int64_t ldb,
+         double beta,
+         DeviceArray& c, int64_t ldc)
+      : GemmBase<T>(opa, opb, m, n, k, alpha, a, lda, b, ldb, beta, c, ldc) {
+    ORT_ENFORCE(alpha == 1.0);
+    ORT_ENFORCE(beta == 0.0);
+
+    if (opa == BlasOp::N && opb == BlasOp::N) {
+      impls_ = this->GetInstances<DeviceGemm<Row, Row>>();
+    } else if (opa == BlasOp::T && opb == BlasOp::N) {
+      impls_ = this->GetInstances<DeviceGemm<Col, Row>>();
+    } else if (opa == BlasOp::N && opb == BlasOp::T) {
+      impls_ = this->GetInstances<DeviceGemm<Row, Col>>();
+    } else if (opa == BlasOp::T && opb == BlasOp::T) {
+      impls_ = this->GetInstances<DeviceGemm<Col, Col>>();
+    }
+
+    ORT_ENFORCE(!impls_.empty());
+    selected_impl_ = 0;
+    UpdateArgumentAndInvoker();
+  }
+
+  std::vector<std::string> ListImpls() const override {
+    std::vector<std::string> results;
+    std::transform(impls_.cbegin(), impls_.cend(), std::back_inserter(results),
+                   [](const auto& it) { return it->GetTypeString(); });
+    return results;
+  }
+
+  bool SelectImpl(std::string name) override {
+    for (int i = 0; i < impls_.size(); i++) {
+      if (impls_[i]->GetTypeString() == name) {
+        selected_impl_ = i;
+        auto is_suppotred = UpdateArgumentAndInvoker();
+        // Run it once before the profiling run, in case there are implicit workspace allocation or some other
+        // allocation during first time launch.
+        if (is_suppotred) {
+          Run();
+        }
+        return is_suppotred;
+      }
+    }
+
+    ORT_THROW("Cannot find implementation ", name);
+  }
+
+  void Run() {
+    dispatch_->Run(arg_.get());
+  }
+
+ private:
+  std::vector<std::unique_ptr<ck::tensor_operation::device::BaseOperator>> impls_;
+  size_t selected_impl_{};
```

Comment:
it is a small integer, maybe just use int?

Reply:
I don't think this matters. This is not in the most inner lool of some high performance kernel anyway. Also, it might cause casting warning if we use `int`.

Reply:
I think see potential casting needed if we use int, which actually removes a casting at line 146: selected_impl_ = i;, noting that i is of type int.

Reply:
There will be change of signedness from `int` to `size_t`. I will switch to also use `size_t` at line 146.

Note, casting is basically unavoidable in std container. There should have been a `ssize_t`...